Jacobhylen/flashing upython firmware guide (#654)
arduino/docs-content: content/tutorials/generic/micropython-firmware/micropython-firmware.md

---
title: 'Flashing MicroPython Firmware to your Arduino Board'
description: 'A guide for flashing the MicroPython firmware to your Arduino Nano 33 BLE Sense, Arduino RP2040 Connect, or Portenta H7, to let you program your board in the MicroPython language.'
tags: 
  - MicroPython
difficulty: Intermediate
libraries:
hardware:
  - hardware/03.nano/boards/nano-33-ble
  - hardware/03.nano/boards/nano-rp2040-connect
  - hardware/04.pro/boards/portenta-h7
software:
  - Arduino Lab for MicroPython
author: "Jacob Hylén"
---
### Introduction

> For some time you have been able to program certain Arduino boards using the OpenMV IDE, writing programs in MicroPython to be run on your board, giving you access to powerful tools such as machine learning with Edge Impulse. But with the release of the (still experimental) Arduino Lab for MicroPython, you have another option for programming your boards in MicroPython! 

Reviewer: The sentence is a bit long, it could be separated in 2/3 sentences if you want

> For some time you have been able to program certain Arduino boards using the OpenMV IDE, writing programs in MicroPython to be run on your board, giving you access to powerful tools such as machine learning with Edge Impulse. But with the release of the (still experimental) Arduino Lab for MicroPython, you have another option for programming your boards in MicroPython! 

Reviewer 2: ```suggestion
For some time you have been able to program selected Arduino boards using the OpenMV IDE and MicroPython. This gives you easy access to powerful technologies such as machine vision and machine learning. 
With the release of the official MicroPython firmwares for the compatible Arduino boards, you have another option for programming your boards in MicroPython for when you don't need machine vision!
To run this firmware you can use the (experimental) Arduino Lab for MicroPython editor. This tutorial explains how to get started.
```

> To do this, you will need to flash a MicroPython firmware to your board that is specifically developed for each board

Reviewer 2: ```suggestion
To do this, you will need to flash a MicroPython firmware to your board that is specifically developed for each board.
```

### Goals

> At the end of this guide you will have flashed a new firmware to your Arduino board, letting you connect to it with the Arduino Lab for MicroPython 

Reviewer 2: ```suggestion
At the end of this guide you will have flashed a new firmware to your Arduino board, letting you connect to it with the Arduino Lab for MicroPython. 
```

#### MacOS
The file you will need to access can be found in `Users/{your-user}/Library/Arduino15/packages/arduino/tools/bossac/1.9.1-arduino2`, the file you are after is `bossac.exe`

>**Note:** The `"Library"` directory is hidden, press `Shift + Command + . ` to reveal hidden directories. 

Open a terminal window and drag and drop the bossac.exe file into the terminal window.

Now, you should be able to execute the command that will flash your board with the MicroPython firmware.

Execute the following command, but replace the port with the name of the port where your board is connected, and the firmware file with the directory where you have placed the firmware you downloaded previously:

```
bossac -e -w --offset=0x16000 --port={port} -i -d -U -R {firmware}
```

Again, don't disconnect or power off your board during this part of the process, watch the progress of flashing the firmware to your board. Once complete, you're all set and you're ready to start programming the board in MicroPython.

Go to the Arduino Lab MicroPython IDE and press connect in the top left, choose the port, and code away!

### Arduino Nano RP2040 Connect

> If you are want to program your Arduino board using MicroPython, and have an Arduino Nano RP2040 to use for this purpose, congratulations! You have chosen the easiest board to flash with the MicroPython firmware!

Reviewer 2: ```suggestion
If you want to program your Arduino board using MicroPython, and have an Arduino Nano RP2040 to use for this purpose, congratulations! You have chosen the easiest board to flash with the MicroPython firmware!
```

All you need to do is to:
download the .uf2 firmware file from the MicroPython [documentation website](docs.arduino.cc/MicroPython), 

> Connect your board to the computer and double tap the reset button to put it in bootloader mode,

Reviewer: Lower case after comma

> Connect your board to the computer and double tap the reset button to put it in bootloader mode,

Reviewer 2: ```suggestion
connect your board to the computer and double tap the reset button to put it in bootloader mode,
```
